src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+    }
+    int sites = width / site_width;
+    const int start_site_index = snapToRowSite(row, start);
+
+    debugPrint(logger_,
+               utl::PAD,
+               "Fill",
+               1,
+               "Filling {} ({:.3f}um, {:.3f}um) -> ({:.3f}um, {:.3f}um)",
+               row->getName(),
+               new_rect.xMin() / dbus,
+               new_rect.yMin() / dbus,
+               new_rect.xMax() / dbus,
+               new_rect.yMax() / dbus);
+    debugPrint(logger_,
+               utl::PAD,
+               "Fill",
+               2,
+               "  start index {} width {}",
+               start_site_index,
+               sites);
+
+    int site_offset = 0;
+    for (auto* filler : fillers) {
+      const int fill_width = filler->getWidth() / site_width;
+      while (fill_width <= sites) {
+        debugPrint(logger_,
+                   utl::PAD,
+                   "Fill",
+                   2,
+                   "    fill cell {} width {} remaining sites {}",
+                   filler->getName(),
+                   fill_width,
+                   sites);
+
+        const std::string name = fmt::format(
+            "IO_FILL_{}_{}_{}", row->getName(), fill_group, site_offset);
+        auto* fill_inst = odb::dbInst::create(block, filler, name.c_str());
+
+        placeInstance(row,
+                      start_site_index + site_offset,
+                      fill_inst,
+                      odb::dbOrientType::R0);
+
+        site_offset += fill_width;
+        sites -= fill_width;
+      }
+    }
+    fill_group++;
+  }
+}
+
+void ICeWall::removeFiller(odb::dbRow* row)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (row == nullptr) {
+    return;
+  }
+  const std::string prefix = fmt::format("IO_FILL_{}_", row->getName());
+
+  for (auto* inst : block->getInsts()) {
+    const std::string name = inst->getName();
+    if (name.substr(0, prefix.length()) == prefix) {
+      odb::dbInst::destroy(inst);
+    }
+  }
+}
+
+void ICeWall::placeBondPads(odb::dbMaster* bond,
+                            const std::vector<odb::dbInst*>& pads,
+                            odb::dbOrientType rotation,
+                            const odb::Point& offset,
+                            const std::string& prefix)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (bond == nullptr) {
+    return;
+  }
+
+  odb::dbMTerm* bond_pin = nullptr;
+  odb::Rect bond_rect;
+  odb::dbTechLayer* bond_layer = nullptr;
+  for (auto* mterm : bond->getMTerms()) {
+    for (auto* mpin : mterm->getMPins()) {
+      for (auto* geom : mpin->getGeometry()) {
+        auto* pin_layer = geom->getTechLayer();
+        if (pin_layer == nullptr) {
+          continue;
+        }
+        if (pin_layer->getRoutingLevel() == 0) {
+          continue;
+        }
+        bond_layer = pin_layer;
+        bond_pin = mterm;
+        bond_rect = geom->getBox();
```

Comment:
The assumption seems to be that there is a single shape.  Is that always true?

Reply:
For the COVER cells, I think so (atleast I have yet to encounter that not being the case)

Reply:
Please add a warning if this is not true